File: Backend/routers/sos.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from datetime import datetime
from database import get_db
from models import User, SOSAlert, LiveLocation, EmergencyContact, SOSStatus
from schemas import (
    SOSTriggerRequest,
    SOSAlertResponse,
    LocationPingRequest,
    SOSStatusResponse,
    LiveLocationResponse
)
from auth import get_current_user
from services.notifications import notify_contacts

router = APIRouter(prefix="/sos", tags=["SOS Emergency"])

@router.post("/trigger", response_model=SOSAlertResponse, status_code=status.HTTP_201_CREATED)
def trigger_sos(
    payload: SOSTriggerRequest,
    db: Session = Depends(get_db)
):
    """
    Trigger an Emergency SOS Distress Alert.

    Academic & Technical Project Flow:
    -----------------------------------
    1. Input Validation:
       FastAPI automatically parses and validates the incoming JSON body against the SOSTriggerRequest Pydantic schema.
       Coordinates (latitude -90..90, longitude -180..180) and battery percentage (0..100) are pre-validated.

    2. User Existence Verification:
       Queries the 'users' table using SQLAlchemy ORM to verify that payload.user_id exists.
       If the user does not exist, an HTTP 404 Not Found exception is raised immediately.

    3. Database Transaction with Error Handling:
       Inserts a new row into the 'sos_alerts' table with status='active' and the current UTC timestamp.
       Wraps database operations inside a try/except block. In case of any database exception,
       db.rollback() is executed to undo pending changes and maintain ACID atomicity, returning an HTTP 500 error.

    4. Response:
       Returns the newly created alert as an SOSAlertResponse model containing id, status, and triggered_at timestamp
       with HTTP status code 201 Created.
    """
    # Step 1: Extract string representation of user_id for DB lookup
    target_user_id = str(payload.user_id)

    # Step 2: Validate user existence in the database (Returns 404 if user is missing)
    user = db.query(User).filter(User.id == target_user_id).first()
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"User with ID '{target_user_id}' does not exist."
        )

    # Step 3: Execute database insertion wrapped in try/except with rollback on error
    try:
        # Create a new SOSAlert ORM instance
        new_alert = SOSAlert(
            user_id=target_user_id,
            latitude=payload.latitude,
            longitude=payload.longitude,
            battery_percent=payload.battery_percent,
            status=SOSStatus.ACTIVE,
            triggered_at=datetime.utcnow()
        )

        # Stage and commit the new alert to the database
        db.add(new_alert)
        db.commit()
        db.refresh(new_alert)

        # Record initial location ping in live_locations table for real-time tracking history
        initial_loc = LiveLocation(
            alert_id=str(new_alert.id),
            lat=payload.latitude,
            lng=payload.longitude,
            battery_pct=payload.battery_percent,
            timestamp=new_alert.triggered_at
        )
        db.add(initial_loc)
        db.commit()

        # Step 4: Dispatch notifications to emergency contacts (handled gracefully if fail)
        user_contacts = db.query(EmergencyContact).filter(EmergencyContact.user_id == target_user_id).all()
        try:
            notify_contacts(
                alert_id=str(new_alert.id),
                user=user,
                contacts=user_contacts,
                lat=payload.latitude,
                lng=payload.longitude
            )
        except Exception as notify_err:
            print(f"[WARNING] Emergency contact notification failed: {notify_err}")

        # Step 5: Return created SOS alert (FastAPI serializes to SOSAlertResponse with 201 Created)
        return SOSAlertResponse(
            id=str(new_alert.id),
            status=str(new_alert.status.value if hasattr(new_alert.status, 'value') else new_alert.status),
            triggered_at=new_alert.triggered_at
        )

    except HTTPException:
        # Re-raise HTTP exceptions (e.g. 404) directly without altering status code
        raise
    except Exception as e:
        # Roll back transaction to restore database to clean state on failure
        db.rollback()
        # Return 500 Internal Server Error with clear error message
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Database transaction failed while triggering SOS alert: {str(e)}"
        )

@router.post("/{alert_id}/location", response_model=LiveLocationResponse)
def push_location_ping(
    alert_id: str,
    payload: LocationPingRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Push a live GPS location ping for an active SOS alert.
    """
    alert = db.query(SOSAlert).filter(SOSAlert.id == alert_id).first()
    if not alert:
        raise HTTPException(status_code=404, detail="SOS Alert not found")

    if alert.status != "active":
        raise HTTPException(status_code=400, detail=f"Cannot push location to an alert with status '{alert.status}'")

    location_ping = LiveLocation(
        alert_id=alert.id,
        lat=payload.lat,
        lng=payload.lng,
        battery_pct=payload.battery_pct,
        timestamp=datetime.utcnow()
    )
    db.add(location_ping)
    db.commit()
    db.refresh(location_ping)

    return LiveLocationResponse.from_orm(location_ping)

@router.get("/{alert_id}/status", response_model=SOSStatusResponse)
def get_sos_status(
    alert_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Poll an SOS alert's status, user details, latest location, and location history.
    """
    alert = db.query(SOSAlert).filter(SOSAlert.id == alert_id).first()
    if not alert:
        raise HTTPException(status_code=404, detail="SOS Alert not found")

    user = db.query(User).filter(User.id == alert.user_id).first()

    # Query location history ordered chronologically
    locations = db.query(LiveLocation).filter(
        LiveLocation.alert_id == alert_id
    ).order_by(LiveLocation.timestamp.desc()).all()

    latest_loc = LiveLocationResponse.from_orm(locations[0]) if locations else None
    loc_history = [LiveLocationResponse.from_orm(loc) for loc in locations]

    return SOSStatusResponse(
        alert_id=alert.id,
        user_id=alert.user_id,
        user_name=user.name if user else "SafeHer User",
        user_phone=user.phone if user else None,
        status=alert.status,
        triggered_at=alert.triggered_at,
        resolved_at=alert.resolved_at,
        latest_location=latest_loc,
        location_history=loc_history
    )

@router.post("/{alert_id}/cancel", response_model=SOSAlertResponse)
def cancel_sos(
    alert_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Cancel or resolve an SOS alert.
    """
    alert = db.query(SOSAlert).filter(SOSAlert.id == alert_id).first()
    if not alert:
        raise HTTPException(status_code=404, detail="SOS Alert not found")

    alert.status = "resolved"
    alert.resolved_at = datetime.utcnow()
    db.commit()
    db.refresh(alert)

    return SOSAlertResponse(
        id=str(alert.id),
        status=str(alert.status.value if hasattr(alert.status, 'value') else alert.status),
        triggered_at=alert.triggered_at
    )

File: Backend/routes/sos.js
/**
 * routes/sos.js — SafeHer Emergency Distress & Live Tracking Router
 *
 * Endpoints:
 *   POST   /sos/trigger           — Authenticated: Trigger SOS distress alert and notify emergency contacts
 *   POST   /sos/:alertId/location — Authenticated: Push live GPS coordinate ping
 *   POST   /sos/:alertId/cancel   — Authenticated: Cancel and resolve active SOS distress signal
 *   GET    /sos/:alertId/status   — Authenticated: Retrieve SOS status, user details, and location history
 *   GET    /sos/track/:token      — Public: Live location tracking map for guardians/emergency services
 */

import express from 'express';
import crypto from 'crypto';
import { v4 as uuidv4 } from 'uuid';
import { query, get, run } from '../config/database.js';
import { requireAuth } from '../middleware/auth.js';
import { notifyAllContacts } from '../services/smsService.js';

const router = express.Router();

// ─────────────────────────────────────────────────────────────────────────────
// PUBLIC ROUTE — must be registered BEFORE /:alertId routes so Express does
// not try to parse "track" as an alertId param.
// GET /sos/track/:token
// ─────────────────────────────────────────────────────────────────────────────
router.get('/track/:token', async (req, res) => {
  try {
    const { token } = req.params;

    // Look up the alert by tracking_token
    const alert = await get(
      'SELECT * FROM sos_alerts WHERE tracking_token = ?',
      [token]
    );

    if (!alert) {
      return res.status(404).json({ detail: 'Tracking link not found or expired.' });
    }

    // Fetch most recent location ping for this alert
    const latestLocation = await get(
      `SELECT lat, lng, battery_pct, timestamp AS recorded_at
         FROM live_locations
        WHERE alert_id = ?
        ORDER BY timestamp DESC
        LIMIT 1`,
      [alert.id]
    );

    // Fetch user info (name only — no PII leak)
    const user = await get('SELECT name FROM users WHERE id = ?', [alert.user_id]);

    const isActive = alert.status === 'active';

    return res.json({
      alertId:         alert.id,
      userName:        user?.name ?? 'SafeHer User',
      status:          alert.status,
      isActive,
      triggeredAt:     alert.triggered_at,
      endedAt:         alert.ended_at ?? null,
      latestLocation:  latestLocation ?? null   // null if no location pings yet
    });
  } catch (err) {
    console.error('Track lookup error:', err);
    return res.status(500).json({ detail: err.message });
  }
});

// POST /sos/trigger — Trigger an Emergency SOS Alert and notify guardians
router.post('/trigger', requireAuth, async (req, res) => {
  try {
    const user = req.user;
    const { latitude, longitude, battery_percent } = req.body;

    const lat     = typeof latitude       === 'number' ? latitude       : 0.0;
    const lng     = typeof longitude      === 'number' ? longitude      : 0.0;
    const battery = typeof battery_percent === 'number' ? battery_percent : 100;

    const trackingToken = crypto.randomBytes(16).toString('hex');
    const alertId = uuidv4();
    const now     = new Date().toISOString();

    // 1. Record distress alert
    await run(
      `INSERT INTO sos_alerts
         (id, user_id, latitude, longitude, battery_percent, status, tracking_token, triggered_at)
       VALUES (?, ?, ?, ?, ?, 'active', ?, ?)`,
      [alertId, user.id, lat, lng, battery, trackingToken, now]
    );

    // 2. Log initial GPS ping
    await run(
      `INSERT INTO live_locations (id, alert_id, lat, lng, battery_pct, timestamp)
       VALUES (?, ?, ?, ?, ?, ?)`,
      [uuidv4(), alertId, lat, lng, battery, now]
    );

    // 3. Build live tracking link
    const frontendUrl  = process.env.FRONTEND_URL || 'http://localhost:3000';
    const trackingLink = `${frontendUrl}/track/${trackingToken}`;

    // 4. Fetch emergency contacts
    const contacts = await query(
      'SELECT * FROM emergency_contacts WHERE user_id = ?',
      [user.id]
    );

    // 5. Dispatch SMS notifications
    const { notified, failed } = await notifyAllContacts(contacts, trackingLink, user.name, battery);

    console.log(`🚨 SOS triggered by ${user.name} | Alert ${alertId} | Token: ${trackingToken}`);
    console.log(`   📲 Notified: ${notified.length} | Failed: ${failed.length}`);

    return res.status(201).json({
      alertId,
      trackingToken,
      trackingLink,
      status:           'active',
      triggeredAt:      now,
      contactsNotified: notified.length,
      contactsFailed:   failed.length,
      failedDetails:    failed   // empty array if all succeeded
    });
  } catch (err) {
    console.error('Trigger SOS error:', err);
    return res.status(500).json({ detail: `Failed to trigger SOS: ${err.message}` });
  }
});

// ─────────────────────────────────────────────────────────────────────────────
// POST /sos/:alertId/location — Push a live GPS location ping
// ─────────────────────────────────────────────────────────────────────────────
router.post('/:alertId/location', requireAuth, async (req, res) => {
  try {
    const { alertId } = req.params;
    const { lat, lng, batteryPct } = req.body;

    if (lat === undefined || lng === undefined) {
      return res.status(400).json({ detail: 'lat and lng are required' });
    }

    // Verify alert exists AND belongs to the authenticated user
    const alert = await get(
      'SELECT * FROM sos_alerts WHERE id = ? AND user_id = ?',
      [alertId, req.user.id]
    );

    if (!alert) {
      return res.status(404).json({ detail: 'SOS alert not found or does not belong to you.' });
    }

    if (alert.status !== 'active') {
      return res.status(403).json({
        detail: `Cannot push location — alert status is '${alert.status}', not 'active'.`
      });
    }

    const pingId  = uuidv4();
    const now     = new Date().toISOString();
    const battery = batteryPct !== undefined ? batteryPct : 100;

    await run(
      `INSERT INTO live_locations (id, alert_id, lat, lng, battery_pct, timestamp)
       VALUES (?, ?, ?, ?, ?, ?)`,
      [pingId, alertId, lat, lng, battery, now]
    );

    return res.json({ id: pingId, alertId, lat, lng, batteryPct: battery, recordedAt: now });
  } catch (err) {
    console.error('Push location error:', err);
    return res.status(500).json({ detail: err.message });
  }
});

// ─────────────────────────────────────────────────────────────────────────────
// POST /sos/:alertId/cancel — Cancel / resolve an active alert
// ─────────────────────────────────────────────────────────────────────────────
router.post('/:alertId/cancel', requireAuth, async (req, res) => {
  try {
    const { alertId } = req.params;

    // Verify ownership
    const alert = await get(
      'SELECT * FROM sos_alerts WHERE id = ? AND user_id = ?',
      [alertId, req.user.id]
    );

    if (!alert) {
      return res.status(404).json({ detail: 'SOS alert not found or does not belong to you.' });
    }

    const now = new Date().toISOString();

    await run(
      `UPDATE sos_alerts SET status = 'cancelled', ended_at = ? WHERE id = ?`,
      [now, alertId]
    );

    return res.json({
      alertId,
      status:      'cancelled',
      triggeredAt: alert.triggered_at,
      endedAt:     now
    });
  } catch (err) {
    console.error('Cancel SOS error:', err);
    return res.status(500).json({ detail: err.message });
  }
});

// ─────────────────────────────────────────────────────────────────────────────
// GET /sos/:alertId/status — Poll full alert status + location history
// ─────────────────────────────────────────────────────────────────────────────
router.get('/:alertId/status', requireAuth, async (req, res) => {
  try {
    const { alertId } = req.params;

    const alert = await get('SELECT * FROM sos_alerts WHERE id = ?', [alertId]);
    if (!alert) {
      return res.status(404).json({ detail: 'SOS Alert not found' });
    }

    const user = await get('SELECT * FROM users WHERE id = ?', [alert.user_id]);

    const locations = await query(
      'SELECT * FROM live_locations WHERE alert_id = ? ORDER BY timestamp DESC',
      [alertId]
    );

    const latestLocation = locations?.length > 0 ? locations[0] : null;

    return res.json({
      alertId:         alert.id,
      userId:          alert.user_id,
      userName:        user?.name  ?? 'SafeHer User',
      userPhone:       user?.phone ?? null,
      status:          alert.status,
      isActive:        alert.status === 'active',
      trackingToken:   alert.tracking_token,
      triggeredAt:     alert.triggered_at,
      endedAt:         alert.ended_at   ?? null,
      latestLocation,
      locationHistory: locations ?? []
    });
  } catch (err) {
    console.error('Get SOS status error:', err);
    return res.status(500).json({ detail: err.message });
  }
});

export default router;
